apps/condo/domains/meter/tasks/sendVerificationDateReminder.js

```diff
@@ -0,0 +1,245 @@
+const { loadListByChunks } = require('@condo/domains/common/utils/serverSchema')
+const { get, flatten, uniq } = require('lodash')
+const { COUNTRIES, DEFAULT_LOCALE } = require('@condo/domains/common/constants/countries')
+const { METER_VERIFICATION_DATE_REMINDER_TYPE } = require('@condo/domains/notification/constants/constants')
+
+const rightJoin = (heads, edges, joinFn, selectFn) => {
+    return heads.map(head => {
+        return selectFn(head, edges.filter(edge => joinFn(head, edge)))
+    })
+}
+
+const readMetersPage = async ({ context, offset, pageSize, date, searchWindowDaysShift, daysCount }) => {
+    // calc window
+    const startWindowDate = dayjs(date).add(searchWindowDaysShift, 'day').format('YYYY-MM-DD')
+    const endWindowDate = dayjs(date)
+        .add(searchWindowDaysShift + daysCount + 1, 'day')
+        .format('YYYY-MM-DD')
+
+    return await Meter.getAll(
+        context, {
+            nextVerificationDate_gte: startWindowDate,
+            nextVerificationDate_lte: endWindowDate,
+        }, {
+            sortBy: 'id_ASC',
+            first: pageSize,
+            skip: offset,
+        }
+    )
+}
+
+const getMetersConnectedWithResidents = async ({ context, meters }) => {
+    // first step is get service consumers by accountNumbers
+    const accountNumbers = uniq(meters.map(meter => meter.accountNumber))
+    const servicesConsumers = await loadListByChunks({
+        context,
+        list: ServiceConsumer,
+        where: {
+            accountNumber_in: accountNumbers,
+        },
+    })
+
+    // second step is to get all resident ids
+    const residentsIds = uniq(servicesConsumers.map(item => item.resident.id))
+
+    // now let's get residents for the list of ids
+    const residents = await loadListByChunks({
+        context,
+        list: Resident,
+        where: {
+            id_in: residentsIds,
+        },
+    })
+
+    // next step - connect residents to services consumers
+    const servicesConsumerWithConnectedResidents = rightJoin(
+        servicesConsumers,
+        residents,
+        (servicesConsumer, resident) => resident.id === servicesConsumer.resident.id,
+        (servicesConsumer, residents) => ({ servicesConsumer, residents })
+    )
+
+    // next step - connect meters to residents using previously created connection
+    const metersWithServiceConsumers = rightJoin(
+        meters,
+        servicesConsumerWithConnectedResidents,
+        (meter, item) => item.servicesConsumer.accountNumber === meter.accountNumber,
+        (meter, servicesConsumers) => ({ meter, servicesConsumers })
+    )
+        .filter(item => item.servicesConsumers != null && item.servicesConsumers.length > 0)
+
+    return metersWithServiceConsumers
+        .map(meterWithServiceConsumers => {
+            const { meter, servicesConsumers } = meterWithServiceConsumers
+
+            const residents = flatten(
+                servicesConsumers.map(item =>
```

Comment:
Because item.residents is part of servicesConsumers item. This is not a join, it actually transformation of servicesConsumers with some sort of filtering. I'll rewrite it just a bit to make it more readable